core/trino-main/src/main/java/io/trino/operator/output/columnar/ColumnarPagePartitioner.java

```diff
@@ -0,0 +1,539 @@
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.slice.Slice;
+import io.airlift.units.DataSize;
+import io.trino.execution.buffer.OutputBuffer;
+import io.trino.execution.buffer.PagesSerde;
+import io.trino.execution.buffer.PagesSerdeFactory;
+import io.trino.operator.OperatorContext;
+import io.trino.operator.PartitionFunction;
+import io.trino.operator.output.PagePartitioner;
+import io.trino.operator.output.PagePartitionerFactory;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.spi.predicate.NullableValue;
+import io.trino.spi.type.Type;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.IntUnaryOperator;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.trino.execution.buffer.PageSplitterUtil.splitPage;
+import static io.trino.operator.output.PartitionedOutputOperator.PartitionedOutputInfo;
+import static io.trino.spi.block.PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class ColumnarPagePartitioner
+        implements PagePartitioner
+{
+    private static final int COLUMNAR_STRATEGY_COEFFICIENT = 4;
+    private final OutputBuffer outputBuffer;
+    private final Type[] sourceTypes;
+    private final PartitionFunction partitionFunction;
+    private final int[] partitionChannels;
+    @Nullable
+    private final Block[] partitionConstantBlocks; // when null, no constants are present. Only non-null elements are constants
+    private final PagesSerde serde;
+    private final PageBuilder[] pageBuilders;
+    private final boolean replicatesAnyRow;
+    private final int nullChannel; // when >= 0, send the position to every partition if this channel is null
+    private final AtomicLong rowsAdded = new AtomicLong();
+    private final AtomicLong pagesAdded = new AtomicLong();
+    private final OperatorContext operatorContext;
+    private final PositionsAppenderFactory positionsAppenderFactory;
+    private final PositionsAppender[] positionsAppenders;
+
+    private boolean hasAnyRowBeenReplicated;
+
+    public static PagePartitionerFactory factory(PositionsAppenderFactory positionsAppenderFactory)
```

Comment:
Revert 6058b291762278d10438e3eecc6f3b13734bf93b, then we don't need this factory method